homeassistant/components/icloud/__init__.py

```diff
@@ -1 +1,629 @@
+                zone_state_long = zone_state.attributes['longitude']
+                zone_distance = distance(
+                    device.location['latitude'],
+                    device.location['longitude'],
+                    zone_state_lat,
+                    zone_state_long)
+                distances.append(round(zone_distance / 1000, 1))
+
+            if distances:
+                mindistance = min(distances)
+            else:
+                continue
+
+            # Calculate out how long it would take for the device to drive
+            # to the nearest zone at 120 km/h:
+            interval = round(mindistance / 2, 0)
+
+            # Never poll more than once per minute
+            interval = max(interval, 1)
+
+            if interval > 180:
+                # Three hour drive?
+                # This is far enough that they might be flying
+                interval = 30
+
+            if (device.battery_level is not None and
+                    device.battery_level <= 33 and
+                    mindistance > 3):
+                # Low battery - let's check half as often
+                interval = interval * 2
+
+            intervals[device.name] = interval
+
+        return max(
+            int(min(
+                intervals.items(),
+                key=operator.itemgetter(1))[1]),
+            self._max_interval)
+
+    def keep_alive(self, now):
+        """Keep the API alive."""
+        if self.api is None:
+            self.reset_account()
+
+        if self.api is None:
+            return
+
+        if self.api.requires_2fa:
+            try:
+                if self.__trusted_device is None:
+                    self.icloud_need_trusted_device()
+                    return
+
+                if self.__verification_code is None:
+                    self.icloud_need_verification_code()
+                    return
+
+                self.api.authenticate()
+                if self.api.requires_2fa:
+                    raise Exception('Unknown failure')
+
+                self.__trusted_device = None
+                self.__verification_code = None
+            except PyiCloudException as error:
+                _LOGGER.error("Error setting up 2FA: %s", error)
+        else:
+            self.api.authenticate()
+
+        self.update_devices()
+
+    def icloud_trusted_device_callback(self, callback_data):
+        """Handle chosen trusted devices."""
+        self.__trusted_device = int(callback_data.get('trusted_device'))
+        self.__trusted_device = self.api.trusted_devices[self.__trusted_device]
+
+        if not self.api.send_verification_code(self.__trusted_device):
+            _LOGGER.error("Failed to send verification code")
+            self.__trusted_device = None
+            return
+
+        if self._accountname in _CONFIGURING:
+            request_id = _CONFIGURING.pop(self._accountname)
+            configurator = self._hass.components.configurator
```

Comment:
The configurator is deprecated. Please make use of a config flow and config entry instead.

https://developers.home-assistant.io/docs/en/config_entries_index.html

Reply:
I thought about that too, but I am going to make it in a new PR, then rebase this one. I think it would be too messy here.

Reply:
Yeah, that would be very nice, I thought about that too.

But should I make it in a new PR to be less messy, then rebase this one, or directly here ?

Reply:
You can do it in a separate PR if you prefer that. We like small PRs. :+1:

Reply:
Hi !

I have some problem to initiate the config flow, can you take a look at this commit (https://github.com/home-assistant/home-assistant/pull/24053/commits/d2bf926ca19c3a857dcdb94b09f7c72a62607373) to help me a bit ?

I inspired me from AdGuard Home, Hangout, Hue, and HomeKit.

I have to logged error event if I added some at `async_setup_entry`, `async_step_user` ...

And nothing came at the integration panel.
<img width="698" alt="Capture d’écran 2019-07-23 à 13 43 08" src="https://user-images.githubusercontent.com/14821482/61710369-caed0d00-ad51-11e9-8c69-7ba85a549a51.png">
_You can notice the "Search" label not well placed, but it used to most of the time (Safari 12.1.1)_

I think that I am missing a little thing but I don't know what.

Also :
How can I replace `self._hass.config.path('icloud')` in the config flow as I can't access `_hass` ?

Thanks.

Reply:
`hass` can be accessed as `self.hass` in the config flow methods. That is set by the core when instantiating the flow.

If strings are missing in the frontend, that can be due to missing translation file. Run the translation script to generate the default english translation file from the strings.json file that you should create manually in the package:
https://developers.home-assistant.io/docs/en/internationalization_backend_localization.html#translation-strings
https://developers.home-assistant.io/docs/en/internationalization_backend_localization.html#configuration-flow-localization

Also make sure you have registered the config flow class:
https://developers.home-assistant.io/docs/en/data_entry_flow_index.html#flow-handler

Reply:
Thanks for the `hass`

I've created a translation and had registered the config flow [here](https://github.com/home-assistant/home-assistant/commit/d2bf926ca19c3a857dcdb94b09f7c72a62607373#diff-f0effb67666f011696a69aaf31d53594R23), still nothing.

Reply:
Make sure you have activated config flow in manifest.json.

Reply:
Same with `"config_flow": true,` in the manifest